packages/storage/src/deque_store.rs

```diff
@@ -0,0 +1,669 @@
+        if self.start >= self.end {
+            return None;
+        }
+        self.end -= 1;
+        let item = self.storage.get_at(self.end);
+        Some(item)
+    }
+
+    // I implement `nth_back` manually because it is used in the standard library whenever
+    // it wants to skip over elements, but the default implementation repeatedly calls next_back.
+    // because that is very expensive in this case, and the items are just discarded, we wan
+    // do better here.
+    // In practice, this enables cheap paging over the storage by calling:
+    // `deque_store.iter().skip(start).take(length).collect()`
+    fn nth_back(&mut self, n: usize) -> Option<Self::Item> {
+        self.end = self.end.saturating_sub(n as u32);
+        self.next_back()
+    }
+}
+
+// This enables writing `deque_store.iter().skip(n).rev()`
+impl<'a, T, S, Ser> ExactSizeIterator for Iter<'a, T, S, Ser>
+where
+    T: Serialize + DeserializeOwned,
+    S: ReadonlyStorage,
+    Ser: Serde,
+{
+}
+
+#[cfg(test)]
+mod tests {
+    use cosmwasm_std::testing::MockStorage;
+
+    use secret_toolkit_serialization::Json;
+
+    use super::*;
+
+    #[test]
+    fn test_pushs_pops() -> StdResult<()> {
+        let mut storage = MockStorage::new();
+        let mut deque_store = DequeStoreMut::attach_or_create(&mut storage)?;
+        deque_store.push_front(&4)?;
+        deque_store.push_back(&5)?;
+        deque_store.push_front(&3)?;
+        deque_store.push_back(&6)?;
+        deque_store.push_front(&2)?;
+        deque_store.push_back(&7)?;
+        deque_store.push_front(&1)?;
+        deque_store.push_back(&8)?;
+
+        assert_eq!(deque_store.pop_front(), Ok(1));
+        assert_eq!(deque_store.pop_back(), Ok(8));
+        assert_eq!(deque_store.pop_front(), Ok(2));
+        assert_eq!(deque_store.pop_back(), Ok(7));
+        assert_eq!(deque_store.pop_front(), Ok(3));
+        assert_eq!(deque_store.pop_back(), Ok(6));
+        assert_eq!(deque_store.pop_front(), Ok(4));
+        assert_eq!(deque_store.pop_back(), Ok(5));
+        assert!(deque_store.pop_back().is_err());
+        Ok(())
+    }
+
+    #[test]
+    fn test_removes() -> StdResult<()> {
+        let mut storage = MockStorage::new();
+        let mut deque_store = DequeStoreMut::attach_or_create(&mut storage)?;
+        deque_store.push_front(&2)?;
+        deque_store.push_back(&3)?;
+        deque_store.push_back(&4)?;
+        deque_store.push_back(&5)?;
+        deque_store.push_back(&6)?;
+        deque_store.push_front(&1)?;
+
+        assert_eq!(deque_store.remove(3), Ok(4));
+        assert_eq!(deque_store.get_at(4), Ok(6));
+        assert_eq!(deque_store.get_at(3), Ok(5));
+        assert_eq!(deque_store.get_at(2), Ok(3));
+        assert_eq!(deque_store.get_at(1), Ok(2));
+        assert_eq!(deque_store.get_at(0), Ok(1));
+
+        assert_eq!(deque_store.remove(1), Ok(2));
+        assert_eq!(deque_store.get_at(3), Ok(6));
+        assert_eq!(deque_store.get_at(2), Ok(5));
```

Comment:
Nice tests!
After doing the changes that i mentioned to `remove`, please make this deque 8 long, then start by trying to pop at `8` and `9` to see that it fails, then popping `0`, then `len -1`, then `1`, then `len - 2`, and then continue however you like until it's empty. That way I'll be sure that the function is behaving well.

Reply:
Done